### PR TITLE
Shell completion fixes/perf improvements

### PR DESCRIPTION
**Pull request by evelikov**
This series ports some zsh completion fixes and perf. improvements over to bash and vice-versa.

One should be able to get `systemctl start` a bit faster on zsh by implementing `_filter_units_by_properties` , effectively executing `systemctl show` once. I'll leave that for the zsh experts since my zsh is rusty.

Note this should substantially help with https://github.com/systemd/systemd/issues/7185

### CODE REVIEW

**Comment by evelikov** [comment]
Humble ping? @yuwata you seem to have some bash experience - care to look at the first two patches?

**Comment by evelikov** [comment]
I've pushed an updated version which addresses my self-review comments.
Most notable difference is ``systemctl start <tab>`` works again. The missing quotes did break it :-\

**Comment by evelikov** [comment]
The artful testing seems very flaky and the failure is unrelated. Don't have permissions to add the `ci-failure-appears-unrelated` label :-\

**Comment by evelikov** [comment]
Humble ping anyone? @yuwata did you had a couple of minutes to test/skim through the patches?

File: shell-completion/bash/systemctl.in [hunk 1]
```diff
@@ -68,35 +68,35 @@ __filter_units_by_properties () {
 
 __get_all_units      () { { __systemctl $1 list-unit-files; __systemctl $1 list-units --all; } \
```
Comment (yuwata) [inline shell-completion/bash/systemctl.in]: Why do not you add `"$2*"` here?

File: shell-completion/bash/systemctl.in [hunk 2]
```diff
@@ -68,35 +68,35 @@ __filter_units_by_properties () {
...
-__get_enabled_units  () { __systemctl $1 list-unit-files  \
+__get_enabled_units  () { __systemctl $1 list-unit-files "$2*" \
         | { while read -r a b c  ; do [[ $b == "enabled"  ]] && echo " $a"; done; }; }
-__get_disabled_units () { __systemctl $1 list-unit-files  \
+__get_disabled_units () { __systemctl $1 list-unit-files "$2*" \
         | { while read -r a b c  ; do [[ $b == "disabled" ]] && echo " $a"; done; }; }
-__get_masked_units   () { __systemctl $1 list-unit-files  \
+__get_masked_units   () { __systemctl $1 list-unit-files "$2*" \
         | { while read -r a b c  ; do [[ $b == "masked"   ]] && echo " $a"; done; }; }
 __get_all_unit_files () { { __systemctl $1 list-unit-files; } | { while read -r a b; do echo " $a"; done; }; }
```
Comment (yuwata) [inline shell-completion/bash/systemctl.in]: Also here.

File: shell-completion/bash/systemctl.in [hunk 3]
```diff
@@ -68,35 +68,35 @@ __filter_units_by_properties () {
...
         __filter_units_by_property $1 CanStart yes $(
-            __systemctl $1 list-unit-files --state enabled,disabled,static | \
+            __systemctl $1 list-unit-files --state enabled,disabled,static "$2*" | \
                     { while read -r a b; do [[ $a =~ @\. ]] || echo " $a"; done; }
-            __systemctl $1 list-units | \
+            __systemctl $1 list-units "$2*" | \
                     { while read -r a b; do echo " $a"; done; } )
 }
-__get_failed_units   () { __systemctl $1 list-units       \
+__get_failed_units   () { __systemctl $1 list-units      "$2*" \
```
Comment (yuwata) [inline shell-completion/bash/systemctl.in]: I do not like this 'indentation'...

File: shell-completion/bash/systemctl.in (hunk 1 above)
Comment (evelikov) [reply shell-completion/bash/systemctl.in]: There's two reasons why I did not add it:
 - the time difference is negligible
 - the zsh completion does not have it

I'm fine with adding it, although unless you feel strongly I'd do that for zsh and bash on top.

File: shell-completion/bash/systemctl.in (hunk 2 above)
Comment (evelikov) [reply shell-completion/bash/systemctl.in]: My earlier reply applies here as well.

File: shell-completion/bash/systemctl.in (hunk 3 above)
Comment (evelikov) [reply shell-completion/bash/systemctl.in]: File already has funky indentation like this. Sure though - will move the new argument just after list-units.

**Comment by yuwata** [comment]
lgtm. I've tested only bash-completion.

**Comment by yuwata** [comment]
@keszybz Could you see the changes for zsh? I would like to include this in v239.

**Comment by keszybz** [comment]
Yep, both bash and zsh parts look great.

**Comment by evelikov** [comment]
Perfect, thanks for the help everyone.

**Pull request merged**